File: android/app/src/main/java/io/github/cyancity/easyunlocker/data/PairingStore.kt
package io.github.cyancity.easyunlocker.data

import android.content.Context
import org.json.JSONArray
import org.json.JSONObject
import java.io.File

/**
 * 一个网关 = 一台 Broker + 这台手机在那里的设备令牌。
 *
 * 网关只决定「往哪儿发待批准 / 决定」，跟保险库没有关系：
 * 切换网关不会碰 vault.eu1，条目照旧。
 */
data class Pairing(
    val id: String,
    val name: String,
    val url: String,
    val deviceToken: String,
    val createdAt: Long = 0L,
)

class PairingStore(context: Context) {
    private val file = File(context.filesDir, "pairings.json")

    fun load(): List<Pairing> {
        if (!file.exists()) return emptyList()
        return runCatching {
            val arr = JSONArray(file.readText())
            buildList {
                for (i in 0 until arr.length()) {
                    val o = arr.getJSONObject(i)
                    val id = o.optString("id")
                    val url = o.optString("url")
                    if (id.isBlank() || url.isBlank()) continue
                    add(
                        Pairing(
                            id = id,
                            name = o.optString("name"),
                            url = url,
                            deviceToken = o.optString("device_token"),
                            createdAt = o.optLong("created_at"),
                        )
                    )
                }
            }
        }.getOrDefault(emptyList())
    }

    fun save(items: List<Pairing>) {
        val arr = JSONArray()
        items.forEach { p ->
            arr.put(
                JSONObject()
                    .put("id", p.id)
                    .put("name", p.name)
                    .put("url", p.url)
                    .put("device_token", p.deviceToken)
                    .put("created_at", p.createdAt),
            )
        }
        file.writeText(arr.toString())
    }
}

/** 比 URL 用的规范化：忽略大小写和结尾斜杠。 */
fun normalizeGatewayUrl(url: String): String = url.trim().trimEnd('/').lowercase()

File: ios/EasyUnlocker/Data/PairingStore.swift
import Foundation

/// pairings.json —— 与安卓 PairingStore 同一文件形状。
final class PairingStore {
    private let file: URL

    init(dir: URL) {
        file = dir.appendingPathComponent("pairings.json")
    }

    func load() -> [Pairing] {
        guard let raw = try? Data(contentsOf: file),
              let arr = try? JSONSerialization.jsonObject(with: raw) as? [[String: Any]] else {
            return []
        }
        return arr.compactMap { o in
            guard let id = o["id"] as? String, !id.isEmpty,
                  let url = o["url"] as? String, !url.isEmpty else { return nil }
            return Pairing(
                id: id,
                name: o["name"] as? String ?? "",
                url: url,
                deviceToken: o["device_token"] as? String ?? "",
                createdAt: (o["created_at"] as? NSNumber)?.int64Value ?? 0
            )
        }
    }

    func save(_ items: [Pairing]) {
        let arr: [[String: Any]] = items.map { p in
            [
                "id": p.id,
                "name": p.name,
                "url": p.url,
                "device_token": p.deviceToken,
                "created_at": p.createdAt,
            ]
        }
        if let raw = try? JSONSerialization.data(withJSONObject: arr) {
            try? raw.write(to: file, options: .atomic)
        }
    }
}
